File: library/src/main/java/com/mindera/skeletoid/analytics/AnalyticsManager.java
package com.mindera.skeletoid.analytics;

import com.mindera.skeletoid.analytics.appenders.IAnalyticsAppender;
import com.mindera.skeletoid.logs.LOG;

import android.app.Activity;
import android.content.Context;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class AnalyticsManager implements IAnalyticsManager {

    private static final String LOG_TAG = "AnalyticsManager";

    /**
     * List of appenders (it can be improved to an ArrayMap if we want to add the support lib as dependency
     */
    private Map<String, IAnalyticsAppender> mAnalyticsAppenders;

    AnalyticsManager() {
        mAnalyticsAppenders = new HashMap<>();
    }

    @Override
    public Set<String> addAppenders(Context context, List<IAnalyticsAppender> analyticsAppenders) {
        if (analyticsAppenders == null || analyticsAppenders.size() == 0) {
            return new HashSet<>();
        }

        final Set<String> appenderIds = new HashSet<>();

        for (IAnalyticsAppender analyticsAppender : analyticsAppenders) {
            analyticsAppender.enableAppender(context);

            final String analyticsId = analyticsAppender.getAnalyticsId();

            if (mAnalyticsAppenders.containsKey(analyticsId)) {
                IAnalyticsAppender oldAnalyticsAppender = mAnalyticsAppenders.remove(analyticsId);
                oldAnalyticsAppender.disableAppender();
                LOG.e(LOG_TAG, "Replacing Analytics Appender with ID: " + analyticsId);
            }

            appenderIds.add(analyticsId);
            mAnalyticsAppenders.put(analyticsId, analyticsAppender);
        }
        return appenderIds;
    }

    @Override
    public void removeAppenders(Context context, Set<String> analyticsIds) {
        if (analyticsIds == null || mAnalyticsAppenders.isEmpty()) {
            return;
        }

        for (String analyticsId : analyticsIds) {
            final IAnalyticsAppender analyticsAppender = mAnalyticsAppenders.remove(analyticsId);
            if (analyticsAppender != null) {
                analyticsAppender.disableAppender();
            }
        }
    }

    @Override
    public void removeAllAppenders() {
        List<String> appendersKeys = new ArrayList<>(mAnalyticsAppenders.keySet());
        for (String analyticsId : appendersKeys) {
            final IAnalyticsAppender analyticsAppender = mAnalyticsAppenders.remove(analyticsId);
            if (analyticsAppender != null) {
                analyticsAppender.disableAppender();
            }
        }
    }

    @Override
    public void trackEvent(String screenName, Map<String, String> analyticsPayload) {
        for (Map.Entry<String, IAnalyticsAppender> entry : mAnalyticsAppenders.entrySet()) {
            entry.getValue().trackEvent(screenName, analyticsPayload);
        }
    }

    @Override
    public void trackPageHit(String screenName, Map<String, String> analyticsPayload) {
        for (Map.Entry<String, IAnalyticsAppender> entry : mAnalyticsAppenders.entrySet()) {
            entry.getValue().trackPageHit(screenName, analyticsPayload);
        }
    }

    @Override
    public void trackPageHit(Activity activity, String screenName, String screenClassOverride, Map<String, String> analyticsPayload) {
        for (Map.Entry<String, IAnalyticsAppender> entry : mAnalyticsAppenders.entrySet()) {
            entry.getValue().trackPageHit(activity, screenName, screenClassOverride, analyticsPayload);
        }
    }
}

File: base/src/main/java/com/mindera/skeletoid/analytics/AnalyticsManager.kt
package com.mindera.skeletoid.analytics

import android.app.Activity
import android.content.Context
import android.os.Bundle
import com.mindera.skeletoid.analytics.appenders.IAnalyticsAppender
import com.mindera.skeletoid.logs.LOG
import java.util.ArrayList
import java.util.HashMap
import java.util.HashSet

class AnalyticsManager internal constructor() : IAnalyticsManager {

    companion object {
        private const val LOG_TAG = "AnalyticsManager"
    }

    /**
     * List of appenders (it can be improved to an ArrayMap if we want to add the support lib as dependency
     */
    private val analyticsAppenders: MutableMap<String, IAnalyticsAppender> = HashMap()

    override fun addAppenders(
        context: Context,
        analyticsAppenders: List<IAnalyticsAppender>
    ): Set<String> {
        val appenderIds: MutableSet<String> = HashSet()
        for (analyticsAppender in analyticsAppenders) {
            analyticsAppender.enableAppender(context)
            val analyticsId = analyticsAppender.analyticsId
            if (this.analyticsAppenders.containsKey(analyticsId)) {
                val oldAnalyticsAppender = this.analyticsAppenders.remove(analyticsId)
                oldAnalyticsAppender!!.disableAppender()
                LOG.e(LOG_TAG, "Replacing Analytics Appender with ID: $analyticsId")
            }
            appenderIds.add(analyticsId)
            this.analyticsAppenders[analyticsId] = analyticsAppender
        }
        return appenderIds
    }

    override fun removeAppenders(
        context: Context,
        analyticsIds: Set<String>
    ) {
        for (analyticsId in analyticsIds) {
            val analyticsAppender = analyticsAppenders.remove(analyticsId)
            analyticsAppender?.disableAppender()
        }
    }

    override fun removeAllAppenders() {
        val appendersKeys: List<String> = ArrayList(analyticsAppenders.keys)
        for (analyticsId in appendersKeys) {
            val analyticsAppender = analyticsAppenders.remove(analyticsId)
            analyticsAppender?.disableAppender()
        }
    }

    override fun trackEvent(
        eventName: String,
        analyticsPayload: Map<String, Any>
    ) {
        for (appender in analyticsAppenders.values) {
            appender.trackEvent(eventName, analyticsPayload)
        }
    }

    override fun trackEvent(
        eventName: String,
        analyticsPayload: Bundle
    ) {
        for (appender in analyticsAppenders.values) {
            appender.trackEvent(eventName, analyticsPayload)
        }
    }

    override fun trackPageHit(
        activity: Activity,
        screenName: String,
        screenClassOverride: String
    ) {
        for (appender in analyticsAppenders.values) {
            appender.trackPageHit(activity, screenName, screenClassOverride)
        }
    }

    override fun setUserID(userID: String) {
        for (appender in analyticsAppenders.values) {
            appender.setUserId(userID)
        }
    }

    override fun setUserProperty(name: String, value: String) {
        for (appender in analyticsAppenders.values) {
            appender.setUserProperty(name, value)
        }
    }
}